fe/estimator.py

```diff
@@ -53,32 +65,57 @@ def simulate(lamb, box, x0, v0, final_potentials, integrator, equil_steps, prod_
         ctxt.add_observable(obs)
 
     prod_schedule = np.ones(prod_steps)*lamb
-    du_dl_freq = 5
-    full_du_dls = ctxt.multiple_steps(prod_schedule, du_dl_freq)
 
-    du_dl_mean, du_dl_std = np.mean(full_du_dls), np.std(full_du_dls)
+    # run MD, optionally pausing every du_dl_freq steps to extract x snapshot
+    if not get_trajectory:
+        xs = None
+        full_du_dls = ctxt.multiple_steps(prod_schedule, du_dl_interval)
+
+    else:
+        xs = [] # in nanometers
+        full_du_dls = []
+
+        # simulate in x_interval-sized chunks
+        t = 0
+        while t < len(prod_schedule):
+            chunk_size = min(len(prod_schedule), (t + x_interval)) - t
+            full_du_dls.append(ctxt.multiple_steps(prod_schedule[t: (t + chunk_size)], chunk_size))
+            xs.append(ctxt.get_x_t())
+            t += chunk_size
+
+        # lists -> arrays
+        xs = np.array(xs)
+        full_du_dls = np.hstack(full_du_dls)
+
 
     # keep the structure of grads the same as that of final_potentials so we can properly
     # form their vjps.
     grads = []
     for obs in du_dp_obs:
         grads.append(obs.avg_du_dp())
 
-    return (du_dl_mean, du_dl_std), grads
+    result = SimulationResult(xs=xs, du_dls=full_du_dls, du_dps=grads)
+    return result
+
 
+class FreeEnergyModel:
```

Comment:
Note that classes make the FreeEnergyModel mutable, (namedtuple version was immutable), I'm guessing this is to manually set the callback later on?

Reply:
Ahh, this wasn't for mutability, but to make callback optional. I didn't realize that namedtuple already supports default constructor arguments (https://docs.python.org/3/library/collections.html#collections.namedtuple), and forgot that the constraint that this needs to be immutable. Will revert

Reply:
Resolved in https://github.com/proteneer/timemachine/pull/360/commits/19b3bb60f7e4f5aa39b854a62b4c230b6eec05b0